### ISSUE TITLE
Paging problems

### ISSUE THREAD

**Issue by makerimages** [makerimages/OSZin#36]
When doing `kmalloc(64);`'s and `kfree`'s , kmalloc twice,  free the first `kmalloc(64);` that you did, DO A `kmalloc(64);` again, and `kfree` the second `kmalloc(64)`, it triplefaults there,

Problably the heap is at completely the wrong place. Where should it be?
Perhaps @Vild can be of assist?

**Comment by Vild** [comment]
Output for the example code?

**Comment by makerimages** [comment]
allocated at 0xc00000c
allocated at 0xc000058  
(free the first)
allocated at 0xc00000c
(free second) triplefault 

@Vild

**Comment by Vild** [comment]
Spam alot of printf:s and see were it crashes

**Comment by makerimages** [comment]
@Vild It's the second free, without it, it doesn't triplefault :P Or did you mean backtracing the kfree?

**Comment by Vild** [comment]
ya, spam alot of debug printf:s in alloc and free

**Comment by makerimages** [comment]
@Vild it's sth in `chunk_glue`. What QEMU eventually says is `Trying to execute code outside RAM or ROM at 0xff53f000`

**Comment by Vild** [comment]
wut, My kernel doesn't work in qemu currently

**Comment by makerimages** [comment]
@Vild Well, bochs does the same thing but gives nothing detailed in the log file. If you want, I can provide you a bochsout of this bug.

**Comment by makerimages** [comment]
@Vild here is a gist of the bochsout.log https://gist.github.com/makerimages/9fae0be1c7363f653260
Apparently it TFs in the first if somewhere. in the glue.

**Comment by Vild** [comment]
You need to turn on debug output on bochs. Else you won't get any debug output

**Comment by Vild** [comment]
My memory stuff requires atleast 256MB of ram

**Comment by makerimages** [comment]
I'm giving it a GB. Bochs stuff tomorrow

**Comment by makerimages** [comment]
@Vild the above Gist updated with debug.log output.

**Comment by makerimages** [comment]
@Vild ideas?

**Comment by Vild** [comment]
Not at the moment, will try any remember to check it later tonight.

**Comment by makerimages** [comment]
@Vild Allright. If you can't find anything to fix this today then I'm leaving tomorrow for the weekend, so you have time to troubleshoot until Sunday :smile: .

**Comment by Vild** [comment]
Okey, I will also see if I have the same problem   in my kernel. I think I
do

**Comment by makerimages** [comment]
And do you, @Vild ?

**Comment by Vild** [comment]
Oh, sorry I forgot. But I have the same result in mine.

**Comment by makerimages** [comment]
Oh god. You think this is something weird-arse with the JamesM tutorial code? @vild

**Comment by Vild** [comment]
Could be something with my changes idk

**Comment by makerimages** [comment]
Might consider rewriting PMM and VMM (again)

**Comment by makerimages** [comment]
Came down to PMM and VMM implementations being crap

**Comment by makerimages** [comment]
Works

**Comment by Vild** [comment]
Wow 118 days since last activity and NOW you close it :p

**Comment by makerimages** [comment]
@Vild Yeah, I got around to implementing [his implementation](https://github.com/levex/osdev/tree/master/memory), after changing my daily driver machine to Linux :smile:

**Comment by makerimages** [comment]
Levex's way was a bit flaky. redoing this. Hopefully for the last time. Branch is ata.

**Comment by Vild** [comment]
+1
